File: cpp/100_SameTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        vector<TreeNode*> p_nodes = {p};
        vector<TreeNode*> q_nodes = {q};
        
        while (!p_nodes.empty() && !q_nodes.empty()) {
            p = p_nodes.back();
            p_nodes.pop_back();
            
            q = q_nodes.back();
            q_nodes.pop_back();
            
            if (!p && !q) { continue; }
            else if (!p || !q || p->val != q->val) { return false; }
            
            p_nodes.push_back(p->left);
            p_nodes.push_back(p->right);
            
            q_nodes.push_back(q->left);
            q_nodes.push_back(q->right);
        }
        return true;
    }
};

File: java/100_SameTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isSameTree(TreeNode pNode, TreeNode qNode) {
        Queue<TreeNode> pNodes = new LinkedList<TreeNode>(){{ add(pNode); }};
        Queue<TreeNode> qNodes = new LinkedList<TreeNode>(){{ add(qNode); }};
        
        while (!pNodes.isEmpty() && !qNodes.isEmpty()) {
            TreeNode p = pNodes.poll();
            TreeNode q = qNodes.poll();
            
            if (p == null && q == null) { continue; }
            else if (p == null || q == null || p.val != q.val) { return false; }
            
            pNodes.add(p.left);
            pNodes.add(p.right);
            
            qNodes.add(q.left);
            qNodes.add(q.right);
        }
        
        return true;
    }
}
